src/walker.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Pipeline;
+  stepFn?: (step: Step) => Step;
+  commandFn?: (command: Command) => Command;
+}
+
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
```

Comment:
add jsdoc here, please.

Also, is there any chance we can change the output type to a nonproducible that we can use to make sure the mutator only accepts an evaluated graph?

Something like:
```suggestion
type EvaluatedPipeline = Pipeline & { __nonproducible: never };

export async function evaluatePipeline(p: Pipeline): Promise<EvaluatedPipeline> {
```

Reply:
Why is this function a public function though? Would anyone want to call this separately from `walk`? If not, we should make `walk` call it automatically.

Reply:
I thought it was best if a pipeline wasn't mutated in a way that the end-user wasn't aware of it

Reply:
Oh yeah, fair enough, it would make `walk` very unintuitive if it just evaluated conditionals.